File: python/src/a2a_utils/artifacts/text.py
"""Text artifact viewing and minimization."""

from typing import Any


class TextArtifacts:
    """Text artifact operations: viewing and minimization."""

    @staticmethod
    def view(
        text: str,
        *,
        line_start: int | None = None,
        line_end: int | None = None,
        character_start: int | None = None,
        character_end: int | None = None,
        character_limit: int = 50_000,
    ) -> str:
        """View text content with optional line or character range selection.

        Args:
            text: The text to view.
            line_start: Starting line number (1-based, inclusive). None = start.
            line_end: Ending line number (1-based, inclusive). None = end.
            character_start: Starting character index (0-based, inclusive). None = start.
            character_end: Ending character index (0-based, exclusive). None = end.
            character_limit: Maximum output size in characters.

        Returns:
            Filtered text string.

        Raises:
            ValueError: If both line and character selection are provided,
                or if parameters are invalid.
        """
        has_line = line_start is not None or line_end is not None
        has_char = character_start is not None or character_end is not None

        if has_line and has_char:
            raise ValueError("Cannot use both line and character selection")

        if has_char:
            start = character_start if character_start is not None else 0
            end = character_end if character_end is not None else len(text)
            result_text = text[start:end]
        elif has_line:
            lines = text.split("\n")
            total_lines = len(lines)
            start_idx, end_idx = TextArtifacts._parse_line_range(line_start, line_end, total_lines)
            selected_lines = lines[start_idx:end_idx]
            result_text = "\n".join(selected_lines)
        else:
            result_text = text

        if len(result_text) > character_limit:
            raise ValueError(
                f"Selected text ({len(result_text):,} characters) exceeds the maximum "
                f"output size of {character_limit:,} characters. "
                f"Try selecting a smaller range."
            )

        return result_text

    @staticmethod
    def minimize(
        text: str,
        *,
        character_limit: int = 50_000,
        tip: str | None = None,
    ) -> dict[str, Any]:
        """Minimize text content for display.

        If text is <= character_limit chars, return it in full inside "text" key.
        If text is > character_limit chars, show first half and last half with metadata.

        Args:
            text: The text content to minimize.
            character_limit: Character limit above which to minimize.
            tip: Tip to include. Defaults to None (no tip); pass a string to include one.

        Returns:
            Dict with "text" key containing readable content and metadata fields.
        """
        if len(text) <= character_limit:
            return {"text": text}

        half = character_limit // 2
        lines = text.split("\n")
        line_count = len(lines)

        # Find which line the half char falls on for the start
        char_count = 0
        start_end_line = 1
        for i, line in enumerate(lines):
            char_count += len(line) + 1  # +1 for newline
            if char_count >= half:
                start_end_line = i + 1
                break

        # Find which line the end section starts on
        char_count = 0
        end_start_line = line_count
        for i in range(len(lines) - 1, -1, -1):
            char_count += len(lines[i]) + 1
            if char_count >= half:
                end_start_line = i + 1
                break

        omitted_chars = len(text) - (2 * half)

        result: dict[str, Any] = {
            "text": f"{text[:half]}\n\n[... {omitted_chars:,} characters omitted ...]\n\n{text[-half:]}",
            "_total_lines": line_count,
            "_total_characters": len(text),
            "_start_line_range": f"1-{start_end_line}",
            "_end_line_range": f"{end_start_line}-{line_count}",
            "_start_character_range": f"0-{half}",
            "_end_character_range": f"{len(text) - half}-{len(text)}",
        }

        if tip is not None:
            result["_tip"] = tip

        return result

    @staticmethod
    def _parse_line_range(
        line_start: int | None, line_end: int | None, total_lines: int
    ) -> tuple[int, int]:
        """Parse line range parameters.

        Args:
            line_start: Starting line number (1-based, inclusive). None means 1.
            line_end: Ending line number (1-based, inclusive). None means total_lines.
            total_lines: Total number of lines

        Returns:
            Tuple of (start_index, end_index) as 0-based indices

        Raises:
            ValueError: If line numbers are invalid
        """
        # Default values
        start = line_start if line_start is not None else 1
        end = line_end if line_end is not None else total_lines

        # Handle negative line numbers (count from end)
        if start < 0:
            start = total_lines + start + 1
        if end < 0:
            end = total_lines + end + 1

        # Validate range
        if start < 1:
            raise ValueError(f"line_start must be >= 1 (got {start})")
        if end > total_lines:
            raise ValueError(f"line_end ({end}) exceeds total lines ({total_lines})")
        if start > end:
            raise ValueError(f"line_start ({start}) must be <= line_end ({end})")

        # Convert to 0-based indices
        return start - 1, end

File: javascript/src/artifacts/text.ts
// SPDX-FileCopyrightText: 2025-present A2A Net <hello@a2anet.com>
//
// SPDX-License-Identifier: Apache-2.0

/**
 * Text artifact viewing and minimization.
 */

/** Format a number with comma separators (locale-independent, matching Python's `f"{n:,}"`). */
function formatNumber(n: number): string {
    return n.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

/** Text artifact operations: viewing and minimization. */
export class TextArtifacts {
    /**
     * View text content with optional line or character range selection.
     *
     * @param text - The text to view.
     * @param opts.lineStart - Starting line number (1-based, inclusive). null = start.
     * @param opts.lineEnd - Ending line number (1-based, inclusive). null = end.
     * @param opts.characterStart - Starting character index (0-based, inclusive). null = start.
     * @param opts.characterEnd - Ending character index (0-based, exclusive). null = end.
     * @param opts.characterLimit - Maximum output size in characters.
     *
     * @returns Filtered text string.
     *
     * @throws Error if both line and character selection are provided,
     *     or if parameters are invalid.
     */
    static view(
        text: string,
        opts?: {
            lineStart?: number | null;
            lineEnd?: number | null;
            characterStart?: number | null;
            characterEnd?: number | null;
            characterLimit?: number;
        },
    ): string {
        const lineStart = opts?.lineStart ?? null;
        const lineEnd = opts?.lineEnd ?? null;
        const characterStart = opts?.characterStart ?? null;
        const characterEnd = opts?.characterEnd ?? null;
        const characterLimit = opts?.characterLimit ?? 50_000;

        const hasLine = lineStart !== null || lineEnd !== null;
        const hasChar = characterStart !== null || characterEnd !== null;

        if (hasLine && hasChar) {
            throw new Error("Cannot use both line and character selection");
        }

        let resultText: string;

        if (hasChar) {
            const start = characterStart ?? 0;
            const end = characterEnd ?? text.length;
            resultText = text.slice(start, end);
        } else if (hasLine) {
            const lines = text.split("\n");
            const totalLines = lines.length;
            const [startIdx, endIdx] = TextArtifacts.parseLineRange(
                lineStart,
                lineEnd,
                totalLines,
            );
            const selectedLines = lines.slice(startIdx, endIdx);
            resultText = selectedLines.join("\n");
        } else {
            resultText = text;
        }

        if (resultText.length > characterLimit) {
            throw new Error(
                `Selected text (${formatNumber(resultText.length)} characters) exceeds the maximum output size of ${formatNumber(characterLimit)} characters. Try selecting a smaller range.`,
            );
        }

        return resultText;
    }

    /**
     * Minimize text content for display.
     *
     * If text is <= characterLimit chars, return it in full inside "text" key.
     * If text is > characterLimit chars, show first half and last half with metadata.
     *
     * @param text - The text content to minimize.
     * @param opts.characterLimit - Character limit above which to minimize.
     * @param opts.tip - Tip to include. Defaults to null (no tip); pass a string to include one.
     *
     * @returns Object with "text" key containing readable content and metadata fields.
     */
    static minimize(
        text: string,
        opts?: {
            characterLimit?: number;
            tip?: string | null;
        },
    ): Record<string, unknown> {
        const characterLimit = opts?.characterLimit ?? 50_000;
        const tip = opts?.tip ?? null;

        if (text.length <= characterLimit) {
            return { text };
        }

        const half = Math.floor(characterLimit / 2);
        const lines = text.split("\n");
        const lineCount = lines.length;

        // Find which line the half char falls on for the start
        let charCount = 0;
        let startEndLine = 1;
        for (let i = 0; i < lines.length; i++) {
            charCount += lines[i].length + 1; // +1 for newline
            if (charCount >= half) {
                startEndLine = i + 1;
                break;
            }
        }

        // Find which line the end section starts on
        charCount = 0;
        let endStartLine = lineCount;
        for (let i = lines.length - 1; i >= 0; i--) {
            charCount += lines[i].length + 1;
            if (charCount >= half) {
                endStartLine = i + 1;
                break;
            }
        }

        const omittedChars = text.length - 2 * half;

        const result: Record<string, unknown> = {
            text: `${text.slice(0, half)}\n\n[... ${formatNumber(omittedChars)} characters omitted ...]\n\n${text.slice(-half)}`,
            _total_lines: lineCount,
            _total_characters: text.length,
            _start_line_range: `1-${startEndLine}`,
            _end_line_range: `${endStartLine}-${lineCount}`,
            _start_character_range: `0-${half}`,
            _end_character_range: `${text.length - half}-${text.length}`,
        };

        if (tip !== null) {
            result._tip = tip;
        }

        return result;
    }

    /**
     * Parse line range parameters.
     *
     * @param lineStart - Starting line number (1-based, inclusive). null means 1.
     * @param lineEnd - Ending line number (1-based, inclusive). null means totalLines.
     * @param totalLines - Total number of lines.
     *
     * @returns Tuple of [startIndex, endIndex] as 0-based indices.
     *
     * @throws Error if line numbers are invalid.
     */
    private static parseLineRange(
        lineStart: number | null,
        lineEnd: number | null,
        totalLines: number,
    ): [number, number] {
        let start = lineStart ?? 1;
        let end = lineEnd ?? totalLines;

        // Handle negative line numbers (count from end)
        if (start < 0) {
            start = totalLines + start + 1;
        }
        if (end < 0) {
            end = totalLines + end + 1;
        }

        // Validate range
        if (start < 1) {
            throw new Error(`line_start must be >= 1 (got ${start})`);
        }
        if (end > totalLines) {
            throw new Error(`line_end (${end}) exceeds total lines (${totalLines})`);
        }
        if (start > end) {
            throw new Error(`line_start (${start}) must be <= line_end (${end})`);
        }

        // Convert to 0-based indices
        return [start - 1, end];
    }
}
